packages/plugin-ext/src/main/browser/file-system-main.ts

```diff
@@ -54,6 +57,21 @@ export class FileSystemMainImpl implements FileSystemMain, Disposable {
         }
     }
 
+    async $readFile(uriComponents: UriComponents): Promise<string> {
+        const uri = Uri.revive(uriComponents);
+        const resource = this.resourceResolver.resolve(new URI(uri));
```

Comment:
just checking: we don't use `FileSystem` directly because an extension can provide custom file system which hooked up via resource resolver and workspace.fs should respect it?

Reply:
yes it's because you can register new filesystem provider and then read/write files from this given provider.